File: packages/sdk/src/commitment.ts
/**
 * Pedersen Commitment Implementation
 *
 * Cryptographically secure Pedersen commitments on secp256k1.
 *
 * ## Security Properties
 *
 * - **Hiding (Computational)**: Cannot determine value from commitment
 * - **Binding (Computational)**: Cannot open commitment to different value
 * - **Homomorphic**: C(v1) + C(v2) = C(v1 + v2) when blindings sum
 *
 * ## Generator H Construction
 *
 * H is constructed using "nothing-up-my-sleeve" (NUMS) method:
 * - Take a well-known string "SIP-PEDERSEN-GENERATOR-H"
 * - Hash it to get x-coordinate candidate
 * - Iterate until we find a valid curve point
 * - This ensures nobody knows the discrete log of H w.r.t. G
 *
 * @see docs/specs/SIP-SPEC.md Section 3.3 - Pedersen Commitment
 */

import { secp256k1 } from '@noble/curves/secp256k1'
import { sha256 } from '@noble/hashes/sha256'
import { bytesToHex, hexToBytes, randomBytes } from '@noble/hashes/utils'
import type { HexString } from '@sip-protocol/types'
import { ValidationError, CryptoError, ErrorCode } from './errors'
import { isValidHex } from './validation'

// ─── Types ───────────────────────────────────────────────────────────────────

/**
 * A Pedersen commitment with associated blinding factor
 */
export interface PedersenCommitment {
  /**
   * The commitment point C = v*G + r*H (compressed, 33 bytes)
   */
  commitment: HexString

  /**
   * The blinding factor r (32 bytes, secret)
   * Required to open/verify the commitment
   */
  blinding: HexString
}

/**
 * A commitment point without the blinding factor (for public sharing)
 */
export interface CommitmentPoint {
  /**
   * The commitment point (compressed, 33 bytes)
   */
  commitment: HexString
}

// ─── Constants ───────────────────────────────────────────────────────────────

/**
 * Domain separation tag for H generation
 */
const H_DOMAIN = 'SIP-PEDERSEN-GENERATOR-H-v1'

/**
 * The generator G (secp256k1 base point)
 */
const G = secp256k1.ProjectivePoint.BASE

/**
 * The independent generator H (NUMS point)
 *
 * Constructed via hash-to-curve with nothing-up-my-sleeve string.
 * Nobody knows log_G(H), making the commitment binding.
 */
const H = generateH()

/**
 * The curve order (number of points in the group)
 */
const CURVE_ORDER = secp256k1.CURVE.n

// ─── Generator H Construction ────────────────────────────────────────────────

/**
 * Generate the independent generator H using NUMS method
 *
 * This uses a try-and-increment approach:
 * 1. Hash the domain separator to get a candidate x-coordinate
 * 2. Try to lift x to a curve point
 * 3. If it fails, increment counter and retry
 *
 * This is similar to how Zcash generates their Pedersen generators.
 */
function generateH(): typeof G {
  let counter = 0

  while (counter < 256) {
    // Create candidate x-coordinate
    const input = new TextEncoder().encode(`${H_DOMAIN}:${counter}`)
    const hashBytes = sha256(input)

    try {
      // Try to create a point from this x-coordinate (with even y)
      // The '02' prefix indicates compressed point with even y
      const pointBytes = new Uint8Array(33)
      pointBytes[0] = 0x02
      pointBytes.set(hashBytes, 1)

      // This will throw if not a valid point
      const point = secp256k1.ProjectivePoint.fromHex(pointBytes)

      // Ensure point is not identity and not G
      if (!point.equals(secp256k1.ProjectivePoint.ZERO) && !point.equals(G)) {
        return point
      }
    } catch {
      // Not a valid point, try next counter
    }

    counter++
  }

  // This should never happen with a properly chosen domain separator
  throw new CryptoError(
    'Failed to generate H point - this should never happen',
    ErrorCode.CRYPTO_FAILED,
    { context: { domain: H_DOMAIN } }
  )
}

// ─── Core Functions ──────────────────────────────────────────────────────────

/**
 * Create a Pedersen commitment to a value
 *
 * C = v*G + r*H
 *
 * Where:
 * - v = value (the amount being committed)
 * - r = blinding factor (random, keeps value hidden)
 * - G = base generator
 * - H = independent generator (NUMS)
 *
 * @param value - The value to commit to (must be < curve order)
 * @param blinding - Optional blinding factor (random 32 bytes if not provided)
 * @returns The commitment and blinding factor
 *
 * @example
 * ```typescript
 * // Create a commitment to 100 tokens
 * const { commitment, blinding } = commit(100n)
 *
 * // Later, prove the commitment contains 100
 * const valid = verifyOpening(commitment, 100n, blinding)
 * ```
 */
export function commit(
  value: bigint,
  blinding?: Uint8Array,
): PedersenCommitment {
  // Validate value type
  if (typeof value !== 'bigint') {
    throw new ValidationError('must be a bigint', 'value', { received: typeof value })
  }

  // Validate value is in valid range
  if (value < 0n) {
    throw new ValidationError('must be non-negative', 'value')
  }
  if (value >= CURVE_ORDER) {
    throw new ValidationError(
      'must be less than curve order',
      'value',
      { curveOrder: CURVE_ORDER.toString(16) }
    )
  }

  // Generate or use provided blinding factor
  const r = blinding ?? randomBytes(32)
  if (r.length !== 32) {
    throw new ValidationError('must be 32 bytes', 'blinding', { received: r.length })
  }

  // Ensure blinding is in valid range (mod n), and non-zero for valid scalar
  const rScalar = bytesToBigInt(r) % CURVE_ORDER
  if (rScalar === 0n) {
    throw new Error('CRITICAL: Zero blinding scalar after reduction - investigate RNG')
  }

  // C = v*G + r*H
  // Handle edge cases where value or blinding could be zero
  let C: typeof G

  if (value === 0n && rScalar === 0n) {
    // Both zero - use identity point (edge case, shouldn't happen with above fix)
    C = secp256k1.ProjectivePoint.ZERO
  } else if (value === 0n) {
    // Only blinding contributes: C = r*H
    C = H.multiply(rScalar)
  } else if (rScalar === 0n) {
    // Only value contributes: C = v*G (shouldn't happen with above fix)
    C = G.multiply(value)
  } else {
    // Normal case: C = v*G + r*H
    const vG = G.multiply(value)
    const rH = H.multiply(rScalar)
    C = vG.add(rH)
  }

  return {
    commitment: `0x${bytesToHex(C.toRawBytes(true))}` as HexString,
    blinding: `0x${bytesToHex(r)}` as HexString,
  }
}

/**
 * Verify that a commitment opens to a specific value
 *
 * Recomputes C' = v*G + r*H and checks if C' == C
 *
 * @param commitment - The commitment point to verify
 * @param value - The claimed value
 * @param blinding - The blinding factor used
 * @returns true if the commitment opens correctly
 */
export function verifyOpening(
  commitment: HexString,
  value: bigint,
  blinding: HexString,
): boolean {
  try {
    // Handle special case of zero commitment (point at infinity)
    if (commitment === '0x00') {
      // Zero commitment only opens to (0, 0) - but that's not valid with our blinding adjustment
      // Actually, zero point means C = C, so it should verify for 0, 0 blinding
      return value === 0n && blinding === ('0x' + '0'.repeat(64))
    }

    // Parse the commitment point
    const C = secp256k1.ProjectivePoint.fromHex(commitment.slice(2))

    // Recompute expected commitment
    const blindingBytes = hexToBytes(blinding.slice(2))
    const rScalar = bytesToBigInt(blindingBytes) % CURVE_ORDER
    if (rScalar === 0n) {
      throw new Error('CRITICAL: Zero blinding scalar after reduction - investigate RNG')
    }

    // Handle edge cases
    let expected: typeof G
    if (value === 0n) {
      expected = H.multiply(rScalar)
    } else if (rScalar === 0n) {
      expected = G.multiply(value)
    } else {
      const vG = G.multiply(value)
      const rH = H.multiply(rScalar)
      expected = vG.add(rH)
    }

    // Check equality
    return C.equals(expected)
  } catch {
    return false
  }
}

/**
 * Create a commitment to zero with a specific blinding factor
 *
 * C = 0*G + r*H = r*H
 *
 * Useful for creating balance proofs.
 *
 * @param blinding - The blinding factor
 * @returns Commitment to zero
 */
export function commitZero(blinding: Uint8Array): PedersenCommitment {
  return commit(0n, blinding)
}

// ─── Homomorphic Operations ──────────────────────────────────────────────────

/**
 * Add two commitments homomorphically
 *
 * C1 + C2 = (v1*G + r1*H) + (v2*G + r2*H) = (v1+v2)*G + (r1+r2)*H
 *
 * Note: The blinding factors also add. If you need to verify the sum,
 * you must also sum the blinding factors.
 *
 * @param c1 - First commitment point
 * @param c2 - Second commitment point
 * @returns Sum of commitments
 * @throws {ValidationError} If commitments are invalid hex strings
 */
export function addCommitments(
  c1: HexString,
  c2: HexString,
): CommitmentPoint {
  // Validate inputs
  if (!isValidHex(c1)) {
    throw new ValidationError('must be a valid hex string', 'c1')
  }
  if (!isValidHex(c2)) {
    throw new ValidationError('must be a valid hex string', 'c2')
  }

  let point1: typeof G
  let point2: typeof G
  try {
    point1 = secp256k1.ProjectivePoint.fromHex(c1.slice(2))
  } catch {
    throw new ValidationError('must be a valid curve point', 'c1')
  }
  try {
    point2 = secp256k1.ProjectivePoint.fromHex(c2.slice(2))
  } catch {
    throw new ValidationError('must be a valid curve point', 'c2')
  }

  const sum = point1.add(point2)

  return {
    commitment: `0x${bytesToHex(sum.toRawBytes(true))}` as HexString,
  }
}

/**
 * Subtract two commitments homomorphically
 *
 * C1 - C2 = (v1-v2)*G + (r1-r2)*H
 *
 * @param c1 - First commitment point
 * @param c2 - Second commitment point (to subtract)
 * @returns Difference of commitments
 * @throws {ValidationError} If commitments are invalid
 */
export function subtractCommitments(
  c1: HexString,
  c2: HexString,
): CommitmentPoint {
  // Validate inputs
  if (!isValidHex(c1)) {
    throw new ValidationError('must be a valid hex string', 'c1')
  }
  if (!isValidHex(c2)) {
    throw new ValidationError('must be a valid hex string', 'c2')
  }

  let point1: typeof G
  let point2: typeof G
  try {
    point1 = secp256k1.ProjectivePoint.fromHex(c1.slice(2))
  } catch {
    throw new ValidationError('must be a valid curve point', 'c1')
  }
  try {
    point2 = secp256k1.ProjectivePoint.fromHex(c2.slice(2))
  } catch {
    throw new ValidationError('must be a valid curve point', 'c2')
  }

  const diff = point1.subtract(point2)

  // Handle ZERO point (identity element) - can't serialize directly
  if (diff.equals(secp256k1.ProjectivePoint.ZERO)) {
    // Return a special marker for zero commitment
    // This is the point at infinity, represented as all zeros
    return {
      commitment: '0x00' as HexString,
    }
  }

  return {
    commitment: `0x${bytesToHex(diff.toRawBytes(true))}` as HexString,
  }
}

/**
 * Add blinding factors (for use with homomorphic addition)
 *
 * When you add commitments, the result commits to (v1+v2) with
 * blinding (r1+r2). Use this to compute the combined blinding.
 *
 * @param b1 - First blinding factor
 * @param b2 - Second blinding factor
 * @returns Sum of blindings (mod curve order)
 */
export function addBlindings(b1: HexString, b2: HexString): HexString {
  const r1 = bytesToBigInt(hexToBytes(b1.slice(2)))
  const r2 = bytesToBigInt(hexToBytes(b2.slice(2)))

  const sum = (r1 + r2) % CURVE_ORDER
  const sumBytes = bigIntToBytes(sum, 32)

  return `0x${bytesToHex(sumBytes)}` as HexString
}

/**
 * Subtract blinding factors (for use with homomorphic subtraction)
 *
 * @param b1 - First blinding factor
 * @param b2 - Second blinding factor (to subtract)
 * @returns Difference of blindings (mod curve order)
 */
export function subtractBlindings(b1: HexString, b2: HexString): HexString {
  const r1 = bytesToBigInt(hexToBytes(b1.slice(2)))
  const r2 = bytesToBigInt(hexToBytes(b2.slice(2)))

  // Handle underflow with modular arithmetic
  const diff = (r1 - r2 + CURVE_ORDER) % CURVE_ORDER
  const diffBytes = bigIntToBytes(diff, 32)

  return `0x${bytesToHex(diffBytes)}` as HexString
}

// ─── Range Proof Support ─────────────────────────────────────────────────────

/**
 * Get the generators for ZK proof integration
 *
 * Returns the G and H points for use in Noir circuits.
 */
export function getGenerators(): {
  G: { x: HexString; y: HexString }
  H: { x: HexString; y: HexString }
} {
  const gAffine = G.toAffine()
  const hAffine = H.toAffine()

  return {
    G: {
      x: `0x${gAffine.x.toString(16).padStart(64, '0')}` as HexString,
      y: `0x${gAffine.y.toString(16).padStart(64, '0')}` as HexString,
    },
    H: {
      x: `0x${hAffine.x.toString(16).padStart(64, '0')}` as HexString,
      y: `0x${hAffine.y.toString(16).padStart(64, '0')}` as HexString,
    },
  }
}

/**
 * Generate a random blinding factor
 */
export function generateBlinding(): HexString {
  return `0x${bytesToHex(randomBytes(32))}` as HexString
}

// ─── Utility Functions ───────────────────────────────────────────────────────

function bytesToBigInt(bytes: Uint8Array): bigint {
  let result = 0n
  for (const byte of bytes) {
    result = (result << 8n) + BigInt(byte)
  }
  return result
}

function bigIntToBytes(value: bigint, length: number): Uint8Array {
  const bytes = new Uint8Array(length)
  let v = value
  for (let i = length - 1; i >= 0; i--) {
    bytes[i] = Number(v & 0xffn)
    v >>= 8n
  }
  return bytes
}

File: sdks/python/sip_protocol/commitment.py
"""
Pedersen Commitment Implementation for SIP Protocol.

Cryptographically secure Pedersen commitments on secp256k1.

Security Properties:
- Hiding (Computational): Cannot determine value from commitment
- Binding (Computational): Cannot open commitment to different value
- Homomorphic: C(v1) + C(v2) = C(v1 + v2) when blindings sum

Generator H Construction:
H is constructed using "nothing-up-my-sleeve" (NUMS) method to ensure
nobody knows the discrete log of H w.r.t. G.
"""

import hashlib
import secrets
from typing import Tuple, Dict

from coincurve import PublicKey, PrivateKey
from coincurve._libsecp256k1 import ffi, lib

from .types import HexString, PedersenCommitment
from .crypto import hex_to_bytes, bytes_to_hex


# Domain separation tag for H generation
H_DOMAIN = "SIP-PEDERSEN-GENERATOR-H-v1"

# secp256k1 curve order
CURVE_ORDER = 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03BBFD25E8CD0364141


def _generate_h() -> bytes:
    """
    Generate the independent generator H using NUMS method.

    Uses try-and-increment approach:
    1. Hash the domain separator to get a candidate x-coordinate
    2. Try to lift x to a curve point
    3. If it fails, increment counter and retry
    """
    for counter in range(256):
        # Create candidate x-coordinate
        input_data = f"{H_DOMAIN}:{counter}".encode("utf-8")
        hash_bytes = hashlib.sha256(input_data).digest()

        try:
            # Try to create a point from this x-coordinate (with even y)
            # The '02' prefix indicates compressed point with even y
            point_bytes = bytes([0x02]) + hash_bytes

            # Validate this is a valid curve point
            PublicKey(point_bytes)

            # Check it's not the identity or base point
            # (coincurve doesn't easily expose G, but this hash won't collide)
            return point_bytes
        except Exception:
            # Not a valid point, try next counter
            continue

    raise RuntimeError("Failed to generate H point - this should never happen")


# The independent generator H (NUMS point)
_H_BYTES = _generate_h()


def _point_multiply(scalar: int, point_bytes: bytes) -> bytes:
    """Multiply a point by a scalar."""
    # Use coincurve for point operations
    point = PublicKey(point_bytes)
    result = point.multiply(scalar.to_bytes(32, "big"))
    return result.format(compressed=True)


def _point_add(point1_bytes: bytes, point2_bytes: bytes) -> bytes:
    """Add two curve points."""
    point1 = PublicKey(point1_bytes)
    result = point1.combine([PublicKey(point2_bytes)])
    return result.format(compressed=True)


def _get_generator_g() -> bytes:
    """Get the base generator G."""
    # G is the public key for private key = 1
    priv = PrivateKey(b"\x00" * 31 + b"\x01")
    return priv.public_key.format(compressed=True)


_G_BYTES = _get_generator_g()


def commit(value: int, blinding: bytes = None) -> Tuple[HexString, HexString]:
    """
    Create a Pedersen commitment to a value.

    C = v*G + r*H

    Where:
    - v = value (the amount being committed)
    - r = blinding factor (random, keeps value hidden)
    - G = base generator
    - H = independent generator (NUMS)

    Args:
        value: The value to commit to (must be < curve order)
        blinding: Optional blinding factor (random 32 bytes if not provided)

    Returns:
        Tuple of (commitment, blinding) as hex strings

    Example:
        >>> commitment, blinding = commit(100)
        >>> verify_opening(commitment, 100, blinding)
        True
    """
    if value < 0:
        raise ValueError("Value must be non-negative")
    if value >= CURVE_ORDER:
        raise ValueError("Value must be less than curve order")

    # Generate or use provided blinding factor
    if blinding is None:
        blinding = secrets.token_bytes(32)
    elif len(blinding) != 32:
        raise ValueError("Blinding must be 32 bytes")

    # Ensure blinding is in valid range (mod n)
    r_scalar = int.from_bytes(blinding, "big") % CURVE_ORDER
    if r_scalar == 0:
        raise RuntimeError("CRITICAL: Zero blinding scalar - investigate RNG")

    # C = v*G + r*H
    if value == 0:
        # Only blinding contributes: C = r*H
        c_bytes = _point_multiply(r_scalar, _H_BYTES)
    else:
        # Normal case: C = v*G + r*H
        v_g = _point_multiply(value, _G_BYTES)
        r_h = _point_multiply(r_scalar, _H_BYTES)
        c_bytes = _point_add(v_g, r_h)

    return (bytes_to_hex(c_bytes), bytes_to_hex(blinding))


def verify_opening(commitment: HexString, value: int, blinding: HexString) -> bool:
    """
    Verify that a commitment opens to a specific value.

    Recomputes C' = v*G + r*H and checks if C' == C

    Args:
        commitment: The commitment point to verify
        value: The claimed value
        blinding: The blinding factor used

    Returns:
        True if the commitment opens correctly
    """
    try:
        # Parse the commitment point
        c_bytes = hex_to_bytes(commitment)

        # Recompute expected commitment
        blinding_bytes = hex_to_bytes(blinding)
        r_scalar = int.from_bytes(blinding_bytes, "big") % CURVE_ORDER

        if value == 0:
            expected = _point_multiply(r_scalar, _H_BYTES)
        else:
            v_g = _point_multiply(value, _G_BYTES)
            r_h = _point_multiply(r_scalar, _H_BYTES)
            expected = _point_add(v_g, r_h)

        return c_bytes == expected
    except Exception:
        return False


def commit_zero(blinding: bytes) -> Tuple[HexString, HexString]:
    """
    Create a commitment to zero with a specific blinding factor.

    C = 0*G + r*H = r*H

    Useful for creating balance proofs.

    Args:
        blinding: The blinding factor

    Returns:
        Tuple of (commitment, blinding)
    """
    return commit(0, blinding)


def add_commitments(c1: HexString, c2: HexString) -> HexString:
    """
    Add two commitments homomorphically.

    C1 + C2 = (v1*G + r1*H) + (v2*G + r2*H) = (v1+v2)*G + (r1+r2)*H

    Note: The blinding factors also add. If you need to verify the sum,
    you must also sum the blinding factors.

    Args:
        c1: First commitment point
        c2: Second commitment point

    Returns:
        Sum of commitments
    """
    c1_bytes = hex_to_bytes(c1)
    c2_bytes = hex_to_bytes(c2)
    result = _point_add(c1_bytes, c2_bytes)
    return bytes_to_hex(result)


def subtract_commitments(c1: HexString, c2: HexString) -> HexString:
    """
    Subtract two commitments homomorphically.

    C1 - C2 = (v1-v2)*G + (r1-r2)*H

    Args:
        c1: First commitment point
        c2: Second commitment point (to subtract)

    Returns:
        Difference of commitments
    """
    c1_bytes = hex_to_bytes(c1)
    c2_bytes = hex_to_bytes(c2)

    # Negate c2 by negating the y-coordinate (flip parity byte)
    c2_negated = bytes([0x03 if c2_bytes[0] == 0x02 else 0x02]) + c2_bytes[1:]

    result = _point_add(c1_bytes, c2_negated)
    return bytes_to_hex(result)


def add_blindings(b1: HexString, b2: HexString) -> HexString:
    """
    Add blinding factors (for use with homomorphic addition).

    Args:
        b1: First blinding factor
        b2: Second blinding factor

    Returns:
        Sum of blindings (mod curve order)
    """
    r1 = int.from_bytes(hex_to_bytes(b1), "big")
    r2 = int.from_bytes(hex_to_bytes(b2), "big")
    result = (r1 + r2) % CURVE_ORDER
    return bytes_to_hex(result.to_bytes(32, "big"))


def subtract_blindings(b1: HexString, b2: HexString) -> HexString:
    """
    Subtract blinding factors (for use with homomorphic subtraction).

    Args:
        b1: First blinding factor
        b2: Second blinding factor (to subtract)

    Returns:
        Difference of blindings (mod curve order)
    """
    r1 = int.from_bytes(hex_to_bytes(b1), "big")
    r2 = int.from_bytes(hex_to_bytes(b2), "big")
    result = (r1 - r2 + CURVE_ORDER) % CURVE_ORDER
    return bytes_to_hex(result.to_bytes(32, "big"))


def generate_blinding() -> HexString:
    """Generate a random blinding factor."""
    return bytes_to_hex(secrets.token_bytes(32))


def get_generators() -> Dict[str, Dict[str, HexString]]:
    """
    Get the generators for ZK proof integration.

    Returns the G and H points for use in ZK circuits.
    """
    g_point = PublicKey(_G_BYTES)
    h_point = PublicKey(_H_BYTES)

    # Get uncompressed format to extract x, y coordinates
    g_uncompressed = g_point.format(compressed=False)
    h_uncompressed = h_point.format(compressed=False)

    return {
        "G": {
            "x": bytes_to_hex(g_uncompressed[1:33]),
            "y": bytes_to_hex(g_uncompressed[33:]),
        },
        "H": {
            "x": bytes_to_hex(h_uncompressed[1:33]),
            "y": bytes_to_hex(h_uncompressed[33:]),
        },
    }
